File: Python/733. Flood Fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        if image[sr][sc] == newColor: return image
        w, h = len(image), len(image[0])
        origin = image[sr][sc]
        def dfs(x, y):
            if not (0 <= x < w) or not (0 <= y < h):
                return image
            if image[x][y] == origin:
                image[x][y] = newColor
                dfs(x - 1, y)
                dfs(x + 1, y)
                dfs(x, y - 1)
                dfs(x, y + 1)
        
        dfs(sr, sc)
        return image

File: TypeScript/733. Flood Fill.ts
function floodFill(image: number[][], sr: number, sc: number, color: number): number[][] {
    var o: number = image[sr][sc], ir: number = image.length, ic: number = image[0].length;
    if(o == color) return image;
    const dfs = (r, c) => {
        if(image[r][c] == o){
            image[r][c] = color;
            if(r > 0) dfs(r - 1, c);
            if(r + 1 < ir) dfs(r + 1, c);
            if(c > 0) dfs(r, c - 1);
            if(c + 1 < ic) dfs(r, c + 1);
        }
    }
    dfs(sr, sc);
    return image;
};
